src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.cs

```diff
@@ -374,6 +289,62 @@ void RecordClosureScope(Scope innermost, Scope outermost, Closure closure)
             }
 
             /// <summary>
+            /// Walk up the scope tree looking for a closure.
+            /// </summary>
+            /// <returns>
+            /// A tuple of the found <see cref="Closure"/> and the <see cref="Scope"/> it was found in.
+            /// </returns>
+            public static (Closure, Scope) FindClosureInScope(Scope startingScope, MethodSymbol closureSymbol)
+            {
+                var currentScope = startingScope;
+                while (currentScope != null)
+                {
+                    var found = currentScope.Closures.SingleOrDefault(c => c.OriginalMethodSymbol == closureSymbol);
+                    if (found != null)
+                    {
+                        return (found, currentScope);
+                    }
+                    currentScope = currentScope.Parent;
+                }
+                throw ExceptionUtilities.Unreachable;
+            }
+
+            /// <summary>
+            /// Finds a <see cref="Closure"/> with a matching original symbol somewhere in the given scope or nested scopes.
+            /// </summary>
+            public static Closure FindClosureInTree(Scope rootScope, MethodSymbol closureSymbol)
+            {
+                var result = Helper(rootScope);
+                if (result != null)
+                {
+                    return result;
+                }
+                throw ExceptionUtilities.Unreachable;
+
+                Closure Helper(Scope scope)
+                {
+                    foreach (var closure in scope.Closures)
```

Comment:
In the FindClosureInScope you used SingleOrDefault to accomplish the exact same operation. Why use `foreach` here but `SingleOrDefault` there? Seems like these should do the same operation.